hw3

AbylaiNur opened the pull request:
hw3

AbylaiNur committed:
hw3

AbylaiNur committed:
Fixed bot:

 -added dependency injection to all bot commands and UserMessageProcessor

 -removed all of the static methods

 -renamed UserRepository to UserDAO

Added tests:

 -added unit tests to UserMessageProcessor.

Before I couldn't add unit tests because all of the methods for the bot were written in static and had outside dependencies, which was a wrong decision.

Русский:

Исправил бота:

 - добавлено dependency injection ко всем командам бота и UserMessageProcessor

 - удалены все статические методы

 -переименовал UserRepository в UserDAO

Добавил тесты:

 -добавил unit тесты в UserMessageProcessor.

До этого я не смог добавить unit тесты, потому что все методы для бота были написаны в static и классы имели внешние зависимости, что было ужасным решением.

AbylaiNur commented:
Во втором push исправил бота и добавил unit тест ему (в commit message подробнее изложено). Я хотел после review сделать второй push, но подумал что буду тратить ваше время. Это нормально? Или лучше так не делать?

AirEM commented on bot/src/main/java/ru/tinkoff/edu/java/bot/BotApplication.java:
```diff
@@ -3,14 +3,24 @@
+
+        try {
+            TelegramBotsApi telegramBotsApi = new TelegramBotsApi(DefaultBotSession.class);
+            telegramBotsApi.registerBot(ctx.getBean(MyBot.class));
```
Рабочая схема, но можно элегантнее

AirEM commented on bot/src/main/java/ru/tinkoff/edu/java/bot/command/Command.java:
```diff
@@ -0,0 +1,15 @@
+
+@Getter
+@AllArgsConstructor
+public abstract class Command {
```
Почему не интерфейс ?

AbylaiNur replied:
Хотел что бы при наследований поля были не статическими и не константными

AirEM commented on bot/src/test/java/ru/tinkoff/edu/java/bot/UserMessageProcessorTest.java:
```diff
@@ -0,0 +1,95 @@
+import static org.mockito.Mockito.*;
+
+@ExtendWith(MockitoExtension.class)
+public class UserMessageProcessorTest {
```
Если это тест на моках, то стоит в рамках него тестировать только логику внутри UserMessageProcessor. Все остальное мокировать. Если расширить границы тестов, то лучше поднимать контекст.

AbylaiNur replied:
Что вы имеете ввиду под контекст? Я не понял

AirEM commented on bot/src/test/java/ru/tinkoff/edu/java/bot/UserMessageProcessorTest.java:
```diff
@@ -0,0 +1,95 @@
+
+    @BeforeEach
+    public void setUp() {
+        commandsMock.add(listCommandMock);
```
Плюс, я бы предложил тесты на listCommand вынести в отдельный класс. На моках протестировать только логику кнутри listCommand

AirEM commented on link-parser/pom.xml:
```diff
@@ -15,16 +15,6 @@
-        </dependency>
-        <dependency>
-            <groupId>org.junit.jupiter</groupId>
-            <artifactId>junit-jupiter-engine</artifactId>
```
А где тесты на link-parser ?

AbylaiNur replied:
Я их добавил когда делал вторую домашку.
Вот путь:
tinkoff-hw/link-parser/src/test/java/ru/tinkoff/edu/java/linkParser/parsers/

Или вы имеете ввиду что надо тестировать GlobalLinkParser?

AirEM replied:
Все норм, увидел

AirEM commented:
Доп балл докинул, поправишь замечания, доки оставшиеся два

AirEM commented:
> Во втором push исправил бота и добавил unit тест ему (в commit message подробнее изложено). Я хотел после review сделать второй push, но подумал что буду тратить ваше время. Это нормально? Или лучше так не делать?

Не понял о чем ты) Если есть, что пушить, всегда смело пуш. Я разберусь, если будут какие-то вопросы, то обсудим

AbylaiNur committed:
Changes:

 - Moved unit tests for ListCommand from UserMessageProccessorTest

 - Added TelegramBotsApiInitilizer to create the bot after initialization of beans

 - Removed bot initialization in BotApplication

AbylaiNur commented:
https://github.com/AbylaiNur/tinkoff-hw/blob/96e85f100435eaecc1513669fc190199b71e070f/bot/src/main/java/ru/tinkoff/edu/java/bot/command/Command.java#L10-L15
Я тут использовал abstract чтобы при наследовании каждый дочерний класс Command имел command и describtion. String command нужен чтобы знать при в каком вводе от пользователя команда активируется. А String describtion просто описывает команду и плюс показывает описание команды в меню.
